### ISSUE TITLE
Unit Test fails for UWP Packaged version of Window Form (WF) application

### ISSUE THREAD

**Issue by bobbryant** [microsoft/WinAppDriver#890]
The VS solution consists of:
1.	WF Project
2.	UWP Package Project
3.	Unit Test Project
4.	Install Project for WF app

The test project code works for the WF app located in the VS development path and after it is installed on the system by changing the test code AppId string  in the appCapabilities.SetCapability("app", AppId); statement.

For the UWP version I set the test code’s AppId to the value in the RegisteredUserModeAppID in the .\AppX\vs.appxricipe file generated by the UWP Package Project.  

When I start the test the application opens followed by a delay of a few seconds then I get the following error messages from the WinAppDriver and an Exception User Unhandled dialog box that pop up in the test project:

Note: The app and appid strings have been replaced with xxx, processId string has been replaced with yyy, and file path strings have been replaced with zzz

WinAppDriver message: 
{"desiredCapabilities":{"platformName":"Windows","app":"xxx","deviceName":"WindowsPC"},"capabilities":{"firstMatch":[{"platformName":"Windows"}]}}
HTTP/1.1 500 Internal Error
Content-Length: 230
Content-Type: application/json

{"status":13,"value":{"error":"unknown error","message":"Failed to locate opened application window with appId: xxx, and processId: yyy"}}

==============================================================

Exception User-Unhandled dialog box pops up on the test code Line:
session = new WindowsDriver<WindowsElement>(new Uri(WindowsApplicationDriverUrl), appCapabilities); 

System.InvalidOperationException
  HResult=0x80131509
  Message=Failed to locate opened application window with appId: xxx, and processId: yyy
  Source=WebDriver

  StackTrace:
   at OpenQA.Selenium.Remote.RemoteWebDriver.UnpackAndThrowOnError(Response errorResponse)
   at OpenQA.Selenium.Remote.RemoteWebDriver.Execute(String driverCommandToExecute, Dictionary`2 parameters)
   at OpenQA.Selenium.Appium.AppiumDriver`1.Execute(String driverCommandToExecute, Dictionary`2 parameters)
   at OpenQA.Selenium.Remote.RemoteWebDriver.StartSession(ICapabilities desiredCapabilities)
   at OpenQA.Selenium.Remote.RemoteWebDriver..ctor(ICommandExecutor commandExecutor, ICapabilities desiredCapabilities)
   at OpenQA.Selenium.Appium.AppiumDriver`1..ctor(Uri remoteAddress, ICapabilities desiredCapabilities)
   at OpenQA.Selenium.Appium.Windows.WindowsDriver`1..ctor(Uri remoteAddress, DesiredCapabilities desiredCapabilities)
   at BIPNC_UiTests.uiTestSession.Setup(TestContext context) in zzz:line 91
   at BIPNC_UiTests.uiTests.ClassInitilize(TestContext context) in zzz:line 61

Note:
line 61 is          Setup(context);
line 91 is             session = new WindowsDriver<WindowsElement>(new Uri(WindowsApplicationDriverUrl), appCapabilities);

**Comment by BalarajAdappa** [comment]
Hi @bobbryant ,

The answer in my case was to set "Developer Mode". Just go to the Start menu and type "Use Developer Features". Set "Developer mode" to On and agree to the instal etc. 

and run the WinAppDriver.exe in background .
it will fix your issue.

**Comment by bobbryant** [comment]
I have development mode enabled.

 

Bob Bryant

Email address:  <mailto:bob.bryant@pobox.com> bob.bryant@pobox.com

Cell Phone:       602.430.2024

**Comment by hassanuz** [comment]
Hi @bobbryant,

Are you still facing this issue? If so, I believe this has to do with a bug with how WinAppDriver tries to locate some WPF apps. 

A workaround is the following: 

1) Create your dedicated app session as you would normally - but place inside a try - catch expression since the app session is **expected** to fail. At the very least, this would launch your app even though WinAppDriver will be unable to attach onto it. 

2) Create a new Desktop (root) Session. The purpose of this session would be to locate and switch to the WPF app window that the previously failed attempt launched. 

3) Override your dedicated app session with the newly attached one. 

The code will potentially look something like this: 

```C#
        protected static WindowsDriver<WindowsElement> session;
        protected static WindowsDriver<WindowsElement> DesktopSession;
```
```C# 
public static void Setup(TestContext context)
        {
            if (session == null)
            {
                // Create a new session to bring up an instance of the Calculator application
                var appiumOptions = new AppiumOptions();
                appiumOptions.AddAdditionalCapability("app", WPFAPPID);
                appiumOptions.AddAdditionalCapability("deviceName", "WindowsPC");
                DesktopSession = null;
                try
                {
                    Console.WriteLine("Trying to Launch App");
                    DesktopSession = new WindowsDriver<WindowsElement>(new Uri(WindowsApplicationDriverUrl), appiumOptions);
                }
                catch
                {
                    Console.WriteLine("Failed to attach to app session (expected).");
                } 
                 //Create Desktop (root) Session
                appiumOptions.AddAdditionalCapability("app", "Root");
                DesktopSession = new WindowsDriver<WindowsElement>(new Uri(WindowsApplicationDriverUrl), appiumOptions);
                //Use inspect.exe to identify the main app window handle. This is what WinAppDriver will be 
                //latching onto. 
                var mainWindow = DesktopSession.FindElementByAccessibilityId("APP WINDOW HANDLE");
                Console.WriteLine("Getting Window Handle");
                var mainWindowHandle = mainWindow.GetAttribute("NativeWindowHandle");
                mainWindowHandle = (int.Parse(mainWindowHandle)).ToString("x"); // Convert to Hex
                appiumOptions = new AppiumOptions();
                appiumOptions.AddAdditionalCapability("appTopLevelWindow", mainWindowHandle);
                session = new WindowsDriver<WindowsElement>(new Uri(WindowsApplicationDriverUrl), appiumOptions);
            }
        }
```

**Comment by hassanuz** [comment]
Please re-open or create a new issue if the above wasn't able to address it^.

**Comment by bobbryant** [comment]
Thanks for the response

I know it’s been awhile but I finally got back to work on this problem. Using your example I can now test both the desktop and the UWP versions of my app by just changing the value in WPFAPPID.  The original app is from a Windows form project that I have added a Window Application Packing Project (UWP) and a  Unit Test Project.  At present there is very little difference between the desktop and UWP versions of the app that I need to tested.  

 

I have not found a consistent set of documentation on using appium for testing in C# and most of my knowledge has come from examples on the web.

Thank again

 

Bob Bryant

Email address:  <mailto:bob.bryant@pobox.com> bob.bryant@pobox.com

Cell Phone:       602.430.2024
